lib/utils.js

```diff
@@ -0,0 +1,177 @@
+var zlib                  = require('zlib');
+var xmldom                = require('xmldom');
+var qs                    = require('querystring');
+var xpath                 = require('xpath');
+var url                   = require('url');
+var xtend                 = require('xtend');
+var signers               = require('./signers');
+var constants             = require('./constants');
+
+var algorithmSignatures = {
+  'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256': 'rsa-sha256',
+  'http://www.w3.org/2000/09/xmldsig#rsa-sha1': 'rsa-sha1'
+};
+
+module.exports.parseSamlRequest = function(req, samlRequest, type, options, callback) {
+  function checkSignature(xml, cb) {
+    var opts = {
+      signingCert: options.signingCert, 
+      thumbprints: options.thumbprints,
+      deflate: !!req.query.Signature,
+      relayState: options.relayState
+    };
+
+    try {
+      validateSignature(req, type, xml, opts);
+      return cb(null, xml);
+    } catch (e) {
+      return cb(e);
+    }
+  }
+
+  function shouldValidateSignature(xml, cb) {
+    if (options.signingCert || options.thumprints) { return cb(null, true) };
+
+    if (!options.getCredentials) {
+      return cb(null, false);
+    }
+
+    var issuerNode = xpath.select(constants.ELEMENTS[type].ISSUER_PATH, xml);
+
+    if (!issuerNode || issuerNode.length === 0) {
+      return cb(null, false);
+    }
+
+    var issuer = issuerNode[0].textContent;
+    options.getCredentials(issuer, function (err, credentials) {
+      if (!credentials){
+        return cb(null, false);
+      }
+
+      options.signingCert = credentials.cert;
```

Comment:
It's weird that a function named `shouldValidateSignature` actually modifies data, options in this case.